src/espnow.rs

```diff
@@ -0,0 +1,183 @@
+        match val {
+            0 => SendStatus::SUCCESS,
+            1 => SendStatus::FAIL,
+            _ => panic!("Wrong status code"),
+        }
+    }
+}
+
+pub type PeerInfo = esp_now_peer_info_t;
+
+#[derive(Debug)]
+struct PrivateData;
+
+#[derive(Debug)]
+pub struct EspNowClient(PrivateData);
+
+impl EspNowClient {
+    pub fn new() -> Result<Self, EspError> {
+        // disable modem sleep, otherwise messages queue up and we're not able
+        // to send any esp-now data after a few messages
+        // esp-idf bug report: https://github.com/espressif/esp-idf/issues/7496
+        esp!(unsafe { esp_wifi_set_ps(0) })?;
+
+        info!("Initializing ESP NOW");
+        esp!(unsafe { esp_now_init() })?;
+        Ok(Self(PrivateData))
+    }
+
+    pub fn new_with_callback(
+        callback: impl for<'b, 'c> FnMut(&'b [u8], &'c [u8]) + 'static + Send,
+    ) -> Result<Self, EspError> {
+        let client = Self::new()?;
+
+        client.register_recv_cb(callback)?;
+        Ok(client)
+    }
+
+    pub fn send(&self, peer_addr: [u8; 6], data: &[u8]) -> Result<(), EspError> {
+        esp!(unsafe {
+            esp_idf_sys::esp_now_send(
+                peer_addr.as_ptr() as *const u8,
+                data.as_ptr() as *const u8,
+                data.len() as size_t,
+            )
+        })?;
+        Ok(())
+    }
+
+    pub fn add_peer(&self, peer_info: PeerInfo) -> Result<(), EspError> {
+        esp!(unsafe { esp_now_add_peer(&peer_info) })?;
+        Ok(())
+    }
+
+    pub fn del_peer(&self, peer_addr: [u8; 6]) -> Result<(), EspError> {
+        esp!(unsafe { esp_now_del_peer(&peer_addr as *const u8) })?;
+        Ok(())
+    }
+
+    pub fn mod_peer(&self, peer_info: PeerInfo) -> Result<(), EspError> {
+        esp!(unsafe { esp_now_mod_peer(&peer_info) })?;
+        Ok(())
+    }
+
+    pub fn get_peer(&self, peer_addr: [u8; 6]) -> Result<PeerInfo, EspError> {
+        let mut peer_info = PeerInfo::default();
+        esp!(unsafe {
+            esp_now_get_peer(
+                &peer_addr as *const u8,
+                &mut peer_info as *mut esp_now_peer_info_t,
+            )
+        })?;
+        Ok(peer_info)
+    }
+
+    pub fn peer_exists(peer_addr: [u8; 6]) -> Result<bool, EspError> {
+        Ok(unsafe { esp_now_is_peer_exist(&peer_addr as *const u8) })
+    }
+
+    pub fn get_peers_number(&self) -> Result<(usize, usize), EspError> {
+        let mut num = esp_now_peer_num_t::default();
+        esp!(unsafe { esp_now_get_peer_num(&mut num as *mut esp_now_peer_num_t) })?;
+        Ok((num.total_num as usize, num.encrypt_num as usize))
+    }
+
+    pub fn set_pmk(&self, pmk: &[u8]) -> Result<(), EspError> {
+        esp!(unsafe { esp_now_set_pmk(pmk.as_ptr()) })?;
+        Ok(())
+    }
+
+    pub fn get_verson() -> Result<u32, EspError> {
+        let mut version: u32 = 0;
+        esp!(unsafe { esp_now_get_version(&mut version as *mut u32) })?;
+        Ok(version)
+    }
+
+    pub fn register_recv_cb(
```

Comment:
I say, we either keep this API, or the `new_with_callback` one. Why would we need both? I'm inclined to a single `new_with_callbacks` API that just supplies both sender and receiver callbacks at construction time. This also allows us to no longer use mutexes to protect these.

Reply:
removed it since I think it's quite common to not register any callbacks